File: src/NnCase.Kernels/MatMul.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Text;

namespace NnCase.Kernels
{
    public static partial class DefaultKernels
    {
        public static void MatMul(ReadOnlySpan<float> inputA, ReadOnlySpan<float> inputB, Span<float> output, ReadOnlySpan<float> bias, int aRows, int aCols, int bCols, ValueRange<float> fusedActivation)
        {
            for (int oy = 0; oy < aRows; oy++)
            {
                for (int ox = 0; ox < bCols; ox++)
                {
                    float value = bias[ox];
                    for (int i = 0; i < aCols; i++)
                    {
                        var a = inputA[oy * aCols + i];
                        var b = inputB[i * bCols + ox];
                        value += a * b;
                    }

                    output[oy * bCols + ox] = ApplyActivation(value, fusedActivation);
                }
            }
        }
    }
}

File: src/kernels/cpu/optimized/x86_64/matmul.cpp
/* Copyright 2019-2021 Canaan Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <iostream>
#include <nncase/kernels/cpu/optimized/tensor_compute.h>
#include <nncase/kernels/cpu/reference/tensor_compute.h>
#include <nncase/kernels/kernel_utils.h>
#include <nncase/runtime/runtime_op_utility.h>

using namespace nncase;
using namespace nncase::runtime;
using namespace nncase::kernels;
using namespace nncase::kernels::cpu;
using namespace nncase::kernels::cpu::optimized;

template result<void> optimized::matmul<float>(const float *input_a, const float *input_b, const float *bias, float *output,
    const runtime_shape_t &in_a_shape, const runtime_shape_t &in_a_strides, const runtime_shape_t &in_b_shape,
    const runtime_shape_t &in_b_strides, const runtime_shape_t &out_shape, const runtime_shape_t &out_strides,
    value_range<float> fused_activation) noexcept;

template <typename T>
result<void> optimized::matmul(const T *input_a, const T *input_b, const T *bias, T *output, const runtime_shape_t &in_a_shape,
    const runtime_shape_t &in_a_strides, const runtime_shape_t &in_b_shape, const runtime_shape_t &in_b_strides,
    const runtime_shape_t &out_shape, const runtime_shape_t &out_strides, value_range<float> fused_activation) noexcept
{
    return cpu::reference::matmul(input_a, input_b, bias, output, in_a_shape, in_a_strides, in_b_shape, in_b_strides, out_shape, out_strides,
        fused_activation);
}